torch/distributed/_fsdp/fully_sharded_data_parallel.py

```diff
@@ -0,0 +1,803 @@
+import copy
+from enum import Enum, auto
+import functools
+import traceback
+from typing import (
+    TYPE_CHECKING,
+    Any,
+    Dict,
+    List,
+    Optional,
+    Set,
+    Tuple,
+    Union,
+)
+
+import torch
+from torch.autograd import Variable
+import torch.distributed as dist
+from torch.distributed import ProcessGroup
+import torch.nn as nn
+import torch.nn.functional as F
+from torch.nn.parameter import Parameter
+
+from torch.distributed.distributed_c10d import _get_default_group
+
+from .utils import (
+    _apply_to_tensors,
+)
+
+if TYPE_CHECKING:
+    from collections import OrderedDict  # noqa: F401
+
+
+class TrainingState(Enum):
+    """
+    Simple enum to indicate what state FSDP is in. Used for asserting
+    to make sure APIs are called in the correct state.
+    ..note::
+        BACKWARD_PRE and BACKWARD_POST states are used to ensure we
+        receives backward hooks in the correct order. It is used to catch
+        unexpected order of hooks being called (likely due to our
+        hook registration logic or autograd engine logic changes).
+    """
+
+    IDLE = auto()
+    FORWARD = auto()
+    BACKWARD_PRE = auto()
+    BACKWARD_POST = auto()
+
+
+class FullyShardedDataParallel(nn.Module):
+    """
+    A wrapper for sharding Module parameters across data parallel workers. This
+    is inspired by `Xu et al.`_ as well as the ZeRO Stage 3 from DeepSpeed_.
+    FullyShardedDataParallel is commonly shorten to FSDP.
+    .. _`Xu et al.`: https://arxiv.org/abs/2004.13336
+    .. _DeepSpeed: https://www.deepspeed.ai/
+    Example::
+        import torch
+        from torch.distributed._fsdp import FullyShardedDataParallel as FSDP
+        torch.cuda.set_device(device_id)
+        sharded_module = FSDP(my_module)
+        optim = torch.optim.Adam(sharded_module.parameters(), lr=0.0001)
+        x = sharded_module(x, y=3, z=torch.Tensor([1]))
+        loss = x.sum()
+        loss.backward()
+        optim.step()
+    .. warning::
+        The optimizer must be initialized *after* the module has been wrapped,
+        since FSDP will shard parameters in-place and this will break any
+        previously initialized optimizers.
+    Args:
+        module (nn.Module):
+            module to be wrapped with FSDP.
+        process_group (Optional):
+            process group for sharding
+    """
+
+    def __init__(
+        self,
+        module: nn.Module,
+        process_group: Optional[ProcessGroup] = None,
+    ):
+        super().__init__()
+        self.module = module
+        self.process_group = process_group or _get_default_group()
+        self.rank = self.process_group.rank()
+        self.world_size = self.process_group.size()
+        # device for computation, if module is on GPU, use module.device;
+        # if module is on CPU, use current device;
+        self.compute_device = _get_default_cuda_device(module)
+        self.compute_dtype = next(module.parameters()).dtype
```

Comment:
Not that I expect anyone to do that, but if `module` has not parameters for some reason, this will crash right?

Reply:
good catch!